tests/unit/test_index.py

```diff
@@ -190,3 +194,85 @@ def test_egg_info_matches(egg_info, search_name, expected):
     link = None     # Only used for reporting.
     version = egg_info_matches(egg_info, search_name, link)
     assert version == expected
+
+
+@pytest.mark.parametrize(
+    "url, vcs_scheme",
+    [
+        ("svn+http://pypi.org/something", "svn"),
+        ("git+https://github.com/pypa/pip.git", "git"),
+    ],
+)
+def test_get_html_page_invalid_scheme(caplog, url, vcs_scheme):
+    """`_get_html_page()` Should error an invalid scheme is given.
+
+    Only file:, http:, https:, and ftp: are allowed.
+    """
+    with caplog.at_level(logging.DEBUG):
+        page = _get_html_page(Link(url), session=Mock(PipSession))
+
+    assert page is None
+    assert caplog.record_tuples == [
+        (
+            "pip._internal.index",
+            logging.DEBUG,
+            "Cannot look at {} URL {}".format(vcs_scheme, url),
+        ),
+    ]
+
+
+def test_get_html_page_directory_append_index(tmpdir):
+    """`_get_html_page()` Should append "index.html" to a directory URL.
+    """
+    dirpath = tmpdir.mkdir("something")
+    dir_url = "file:{}".format(urllib_request.pathname2url(dirpath))
```

Comment:
There was a test failure around this because of the normal of leading slashes. You might want to use `urllib_parse` / look at how the real code does it, which would also help with cross-platform file paths. Alternatively, you could "spot check" the URL passed, like making sure it begins with "file:" and ends with "/index.html".

Also, how about parametrizing and including a path that's both a directory and looks like an archive? That would cover one of the "fixes" in this PR. Even if that will change later, it would be good to be able to know from the test that something is changing.

PS - re: test failures, do you know how to run tests for a single file or single test function using tox? That makes it a lot easier to run tests locally.

Reply:
> Re: tox, you can pass additional arguments directly to tox

Yes, exactly. I was going to tell you for your benefit in case you didn't already know..